pkg/reconciler/build/build.go

```diff
@@ -352,6 +370,64 @@ func (c *Reconciler) conditionForPod(pod *corev1.Pod, stepsCompleted []string) c
 	}
 }
 
+func (c *Reconciler) finalize(ctx context.Context, build *buildapi.Build) error {
+	if !slices.Contains(build.GetFinalizers(), BuildFinalizer) {
+		return nil
+	}
+
+	if build.Finished() {
+		keychain, err := c.KeychainFactory.KeychainForSecretRef(ctx, registry.SecretRef{
+			ServiceAccount: build.Spec.ServiceAccountName,
+			Namespace:      build.Namespace,
+		})
+		if err != nil {
+			return err
+		}
+
+		if err := c.RegistryClient.Delete(keychain, build.Status.LatestImage); err != nil {
+			return err
```

Comment:
Returning error might mean that builds get stuck if the finalized can't delete the image due to immutability rules in the registry. I think this should be a best effort cleanup

Reply:
Isn't it fair to keep the build if the desired outcome is not reached? The solution is not to enable the cleanup when the registry forbids objects deletion is used. If I set the cleanup to `true` but the image is still present, it should be considered an error, not a warning message